cmd/notify-mailer/main.go

```diff
@@ -108,119 +109,141 @@ func (m *mailer) run() error {
 	}
 	return nil
 }
 
-// resolveEmailAddresses looks up the id of each recipient to find that
-// account's email addresses, then adds that recipient to a map from address to
-// recipient struct.
-func (m *mailer) resolveEmailAddresses() (emailToRecipientMap, error) {
-	result := make(emailToRecipientMap, len(m.destinations))
-
-	for _, r := range m.destinations {
-		// Get the email address for the reg ID
-		emails, err := emailsForReg(r.id, m.dbMap)
+// resolveEmailAddresses fetches the contact row of each recipient by
+// registration ID then creates a map of email addresses to (a list of)
+// `recipient`s that resolve to that email address.
+func (m *mailer) resolveEmailAddresses() (addressesToRecipientsMap, error) {
+	m.log.Infof("Resolving %d recipient addresses", len(m.recipients))
+	result := make(addressesToRecipientsMap, len(m.recipients))
+	for _, r := range m.recipients {
+		contact, err := getContactForID(r.id, m.dbMap)
 		if err != nil {
 			return nil, err
 		}
 
-		for _, email := range emails {
-			parsedEmail, err := mail.ParseAddress(email)
+		for _, address := range contact {
+			parsedAddress, err := mail.ParseAddress(address)
 			if err != nil {
-				m.log.Errf("unparsable email for reg ID %d : %q", r.id, email)
+				m.log.Errf("While parsing address: %q for ID: %d: %s", address, r.id, err)
 				continue
 			}
-			addr := parsedEmail.Address
-			result[addr] = append(result[addr], r)
+			result[parsedAddress.Address] = append(result[parsedAddress.Address], r)
 		}
 	}
 	return result, nil
 }
 
-// Since the only thing we use from gorp is the SelectOne method on the
-// gorp.DbMap object, we just define an interface with that method
+// dbSelector since the only methods we use from the `gorp.DbMap` object are
+// `SelectOne` and `Execute` the we just define an interface with those methods
 // instead of importing all of gorp. This facilitates mock implementations for
-// unit tests
+// unit tests.
 type dbSelector interface {
 	SelectOne(holder interface{}, query string, args ...interface{}) error
+	Exec(query string, args ...interface{}) (sql.Result, error)
 }
 
-// Finds the email addresses associated with a reg ID
-func emailsForReg(id int, dbMap dbSelector) ([]string, error) {
-	var contact contactJSON
-	err := dbMap.SelectOne(&contact,
-		`SELECT id, contact
-		FROM registrations
-		WHERE contact != 'null' AND id = :id;`,
+// getContactForID queries the database for the contact field associated with
+// the provided registration ID.
+func getContactForID(id int64, dbMap dbSelector) ([]string, error) {
+	// Transaction isolation level READ UNCOMMITTED trades consistency for
+	// performance.
+	_, err := dbMap.Exec(
+		"SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;",
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"while setting transaction isolation level: %s", err)
+	}
+
+	var result queryResult
+	err = dbMap.SelectOne(&result,
```

Comment:
Does this suffer the same problem discussed previously, where there's no guarantee that the `dbMap.Exec` and the `dbMap.SelectOne` will actually occur on the same connection, and therefore the transaction isolation level won't actually affect anything? It doesn't look to me like the dbSelector being passed in is itself a transaction object.

Reply:
I was thinking that executing just before would _mostly_ ensure the same session was used. But yeah, you're right, if it pulls a different connection from the pool we are right back where we started 🤦🏻‍♀️